python/gtsam/tests/test_Cal3Fisheye.py

```diff
@@ -17,6 +17,15 @@
 from gtsam.utils.test_case import GtsamTestCase
 from gtsam.symbol_shorthand import K, L, P
 
+
+def ulp(ftype=np.float64):
+    """
+    Unit in the last place of floating point datatypes
+    """
+    f = np.finfo(ftype)
+    return f.tiny / ftype(1 << f.nmant)
```

Comment:
The goal here is to use the ulp value to test for a possible overflow in the implementation, independent of the floating point datatype.

The value of ulp() for `half float`, `float` and `double` is given by,
```Julia
julia> eps(nextfloat(Float16(0.0)))
Float16(6.0e-8)
julia> eps(nextfloat(Float32(0.0)))
1.0f-45
julia> eps(nextfloat(Float64(0.0)))
5.0e-324
```

The f.tiny provides the smallest value, at which the floating point number can still be represented as normalized value.
Since python 3.9, numpy provides the `numpy.ulp()` function (see https://bugs.python.org/issue39310 or https://www.boost.org/doc/libs/1_77_0/libs/math/doc/html/math_toolkit/next_float/ulp.html). 

In our case, it defines the smallest denormalized value ulp, such that 0+ulp > 0. In simple terms, when numbers are denormalized, calculations are done in fixed-point arithmetic, using the smallest exponent for the mantissa.

This value is used in the test to check whether overflow occurs. In the original implementation, setting x=ulp() and y=0, resulted in an overflow returning a NaN Jacobian.

In terms of the numeric resolution of double float, a value of 1e-20 is a huge number. For half-float, it is a too small number to represent.

The test would not catch any critical overflow condition. (Same for using f.tiny - it does not catch an overflow for computing the Jacobian)

Physically, thinking in terms of taylor expansion, a radius of 1e-20 the Jacobian would be sufficient to be in the paraxial region  - as noted, the implementation then depends on the datatype.

Reply:
Awesome, this is a great explanation.